### PR TITLE
Issue223: second solution

### PR DESCRIPTION
**Pull request by josojo**
Hey,

here is the code for the second solution discussed in the issue #223. The additional complexity is really low and the upper bound for the gas costs for solution submission would be easier to determine and it would be lower, which would allow us to do more trades.

Probably, I favor this second solution.

What is your opinion?

---
testplan: Unit tests.

### CODE REVIEW

File: contracts/EpochTokenLocker.sol [hunk 1]
```diff
@@ -35,6 +35,9 @@ contract EpochTokenLocker {
     // User => Token => BalanceState
     mapping(address => mapping(address => BalanceState)) private balanceStates;
 
+    // user => token => batchId => wasCredidet
```
Comment (bh2smith) [inline contracts/EpochTokenLocker.sol]:
```suggestion
    // user => token => batchId => wasCredited
```

File: contracts/EpochTokenLocker.sol [hunk 2]
```diff
@@ -35,6 +35,9 @@ contract EpochTokenLocker {
...
+    mapping(address => mapping (address => mapping(uint => bool ))) public hasCreditedBalance;
```
Comment (bh2smith) [inline contracts/EpochTokenLocker.sol]:
```suggestion
    mapping(address => mapping (address => mapping(uint => bool))) public hasCreditedBalance;
```

**Review by bh2smith — APPROVED** [review]
This is pretty nice. I would like to do a second pass through this at some point, but it would be really nice if the description of this PR could summarize what is happening here. I will make my best guess after having read the code.

During auction solution submission, if a token is credited (as a result of exchange), then this token cannot be withdrawn until this auction has passed. All valid withdraw requests less than the current batchId can be withdrawn.

My only question now is... what issue does this resolve? I mean what is the real difference between 
```
old - `addBalanceAndPostponeWithdraw`
new - `addBalanceAndBlockWithdrawForThisBatch`
```

Does it simply not aggregate requests?

**Comment by josojo** [comment]
Hey,

>During auction solution submission, if a token is credited (as a result of exchange), then this token cannot be withdrawn until this auction has passed. All valid withdraw requests less than the current batchId can be withdrawn.

Kind of accurate. The valid withdraw requests can be withdrawn, yes, but not in the current batch.

I hoped that the scope of this PR would be clear by referencing issue #223. But yeah, maybe I should have mentioned it again. 

>old - `addBalanceAndPostponeWithdraw`

With this function, the problem was that the postponed withdrawRequest actually modifies the optimization problem, as this postpone withdrawRequest would have an impact on available balances.
With this Pr, it is fixed

File: contracts/EpochTokenLocker.sol [hunk 3]
```diff
@@ -121,12 +133,17 @@ contract EpochTokenLocker {
         return balance;
     }
 
+    function hasValidWithdrawRequest(address user, address token) public view returns(bool) {
+        return balanceStates[user][token].pendingWithdraws.stateIndex < getCurrentBatchId() &&
+            balanceStates[user][token].pendingWithdraws.stateIndex > 0;
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:138]: Should we also check the amount > 0?

File: contracts/EpochTokenLocker.sol [hunk 4]
```diff
@@ -121,12 +133,17 @@ contract EpochTokenLocker {
...
+    }
+
     /**
      * internal functions
      */
-    function addBalanceAndPostponeWithdraw(address user, address token, uint amount) internal {
-        if (balanceStates[user][token].pendingWithdraws.stateIndex < getCurrentBatchId()) {
-            balanceStates[user][token].pendingWithdraws.stateIndex = getCurrentBatchId();
+    function addBalanceAndBlockWithdrawForThisBatch(address user, address token, uint amount) internal {
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:151]: Do we always want to do the two things together, or is it better to have two different methods? The good think about one method is that you cannot "forget" to block the withdraw when you change the balance.

File: contracts/EpochTokenLocker.sol [hunk 5]
```diff
@@ -71,6 +78,11 @@ contract EpochTokenLocker {
             "withdraw was not registered previously"
         );
 
+        require(
+            !hasCreditedBalance[msg.sender][token][getCurrentBatchId()],
+            "withdraw is not possible, due to new credit in this batchId"
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol]: Maybe more clear "Withdraw not possible for token that is traded in the current auction." ?

File: contracts/EpochTokenLocker.sol [hunk 6]
```diff
@@ -35,6 +35,9 @@ contract EpochTokenLocker {
     // User => Token => BalanceState
     mapping(address => mapping(address => BalanceState)) private balanceStates;
 
+    // user => token => batchId => wasCredited
+    mapping(address => mapping (address => mapping(uint => bool))) public hasCreditedBalance;
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:39]: Would it be slightly more efficient to just store `user => token => largest_credited_batchId` ?

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia0.giphy.com%2Fmedia%2FWUq1cg9K7uzHa%2Fgiphy-downsized-medium.gif&token=yay" data-gitmeme-token="yay"></a>

I also like this solution much more. We should probably comment in the SC somewhere why we need this complicated logic (with an attack example).

File: contracts/EpochTokenLocker.sol (hunk 3 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:138]: Hm... I don't think so. If it returns that there is no validWithdrawRequest, as the amount == 0, and we then credit some amount to it, suddenly this account could do the withdraw before a solution is officially accepted.

File: contracts/EpochTokenLocker.sol (hunk 4 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:151]: Right, I also like the point you mentioned!

File: contracts/EpochTokenLocker.sol (hunk 6 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:39]: I think in the current solution, we effectively only store a boolean, and the user, token, batchID are just getting hashed.

This should be more efficient than storing a uint for the batchID

**Comment by bh2smith** [comment]
> We should probably comment in the SC somewhere why we need this complicated logic (with an attack example).

Maybe even make a separate folder with attacks that wont work on our contract.

File: contracts/EpochTokenLocker.sol (hunk 6 above)
Comment (fleupold) [reply contracts/EpochTokenLocker.sol:39]: Did you verify this? I'd assume that if a mapping is the value of a mapping it's represented as one full hash.

**Comment by bh2smith** [comment]
This was already merged and then added to. To we need to merge again?

**Comment by fleupold** [comment]
It's in master, so I think we are ok: https://github.com/gnosis/dex-contracts/blob/master/contracts/EpochTokenLocker.sol#L39

**Pull request merged**